### ISSUE TITLE
local variables should override/extend global ones without error

### ISSUE THREAD

**Issue by ex-nerd** [httpyac/httpyac.github.io#7]
Both postman and the other popular rest plugin for vscode interpret variables from child to parent, rather than simply in order defined in the file.  This is contrary to traditional "code" interpretation order but allows for "global" variables to define patterns that are then extended by queries later in the document, keeping their sections clean.

Additionally, the session maintains variables between runs, so you can experience some unexpected values in your variables.

E.g. take this code:

```
@domain = postman-echo.com
// @path = path0
@baseUrl = https://{{domain}}/{{path}}

###

@path = path1
GET {{baseUrl}}

###

@path = path2
GET {{baseUrl}}
```

- If you try to execute the first `GET` this will result in an error stating `ReferenceError: path is not defined at Object.`
  - _Given the behavior of other rest clients, I would have expected to resolve `https://postman-echo.com/path1`_
- However, if you uncomment that first `@path` assignment and execute the first query again, you will be sent to `https://postman-echo.com/path0`. Not the expected URL, but at least it didn't get an error.
- Now, re-comment that line and run the query again to see that it no longer produces an error (presumably a cached variable state), but also resolves to `https://postman-echo.com/path0`
- Now run the second query and see the request resolves to `https://postman-echo.com/path1`
  - it seems to cache the value from the previous execution of the first query that resolved to `path1`
- Now run the first query and see the request resolves to `https://postman-echo.com/path2`
  - again, it seems to cache the value from the previous execution so the URL points to `path2`

I realize that the caching is how it's possible to chain calls together, so this is more about the inconsistent application of the variables.  "top down" may be more accurate code-wise, but makes less sense if these requests are thought of as cascading templated queries.

**Comment by ex-nerd** [comment]
Yeah, it's the difference between seeing the document as code executed top to bottom, and variables as any other kind of code variable, vs a sort of macro system where the "global variables" exist to help keep the file clean.

Actually, as I think of this, it's less about when the variables are _set_, and more about when they are _interpreted_.

So variables would be assigned as strings (or whatever) top to bottom, but any interpolated values within those strings would be processed at runtime.

**Comment by AnWeber** [comment]
The variables will not be replaced directly in the future. Instead, it is allowed that these refer to other variables that do not yet exist. Only when NodeJS scripts or a request is made will an attempt be made to replace the variables.
The variables are additionally regenerated per region/request. This means that even if all requests in the file are sent using `Send All`, the variable `path` of your example will be set correctly.

**Comment by AnWeber** [comment]
The variables processing has been updated with the last release. Would you please test it.

**Comment by ex-nerd** [comment]
Everything with variable definition order seems to be working now as expected, however it seems that like with https://github.com/AnWeber/vscode-httpyac/issues/82, I'm not able to use `@name`/`@ref` the same as I could before.

After a bit of digging, it looks like maybe this is just a need to update things.  Existing code was using {{refName.response.body.property}} but looks like now it's just {{refName.property}} (and there is no longer a way to access properties like response headers?). This mostly seems like it just needs documentation to be added -- there is very little information about how to use the ref/name/exports functionality.

P.S.  I just noticed I filed this on the wrong project ... sorry about that.

**Comment by AnWeber** [comment]
I just don't understand how your example with `refName.response.body.property` worked before. The [variable created with @name](https://github.com/AnWeber/httpyac/blob/main/src/actions/responseAsVariableAction.ts#L24) is actually always filled with the body of the response, and if [JSON the parsed body](https://github.com/AnWeber/httpyac/blob/main/src/actions/responseAsVariableAction.ts#L7). This is implemented mainly for compatibility with rest-client. Do you possibly have an example code what worked before.

I also saw the problem with the header and that's why the predefined [`response` variable](https://github.com/AnWeber/httpyac/blob/main/src/actions/responseAsVariableAction.ts#L8) was created ([docs](https://httpyac.github.io/guide/scripting.html#access-to-variables)). After calling the request you could store necessary headers in custom variables ([post request script](https://httpyac.github.io/guide/scripting.html#scripting)).
```
GET http://...

@header = {{response.headers['content-type']}}
{{
  exports.header2 = response.headers['content-type'];
}}
###

GET http://...
```

Writing documentation is definitely not my forte. If you have written the variable logic yourself, the whole idea is relatively simple, but for outsiders it is really a black box. I think it would be great if you could just make an issue for the points that need further documentation. I will be happy to help.

**Comment by ex-nerd** [comment]
> just don't understand how your example with `refName.response.body.property` worked before

Maybe it didn't. It's possible I copied that code from the other vscode REST plugin and just never got far enough with httpyac (because of this current issue) to realize it.  Checking those docs, this seems to be the most likely issue, since their examples all use the `response.body` construct.

Anyway, everything seems to be working great now, so I think this ticket is ready to be closed.  I'll look for doc areas that could be expanded, and file a separate ticket for that.  At the moment, it looks like:

- More clear explanation and example(s) of how to use `@name` and `@ref`, since the only place I can find
  mention of these just says _"You can also `@ref`erence `@name`d meta data"_ but doesn't really explain
  what the properties are.
- How to use `response` and post-request scripts with `@ref`

Probably both addressed as a "passing data around" section.

**Comment by AnWeber** [comment]
* I call [@name](https://httpyac.github.io/guide/metaData.html#name) and [@ref](https://httpyac.github.io/guide/metaData.html#ref-and-forceref) as Meta Data. Did you find this documentation before?
* Actually, `response` and `@ref` have nothing to do with each other directly. 
  * `response` is only the variable name to access the current response in post request scripts/ variables (all scripts after request line). It is just a variable. 
  * `@ref` is the instruction that the referenced request should please be executed before. But this way the variables are added to the [run variables](https://github.com/AnWeber/httpyac/blob/main/src/actions/refMetaAction.ts#L23) and thus accessible in the following scripts/ variables

Your suggestion with `passing around data` is a good hint.

**Comment by ex-nerd** [comment]
> * I call [@name](https://httpyac.github.io/guide/metaData.html#name) and [@ref](https://httpyac.github.io/guide/metaData.html#ref-and-forceref) as Meta Data. Did you find this documentation before?

Actually, no.  Google's "site:..." search apparently finds "ref" and "name" too common to turn up the important stuff.  Maybe the only fix needed here is to link from [import variables](https://httpyac.github.io/guide/variables.html#import-variables) to the MetaData page.

And yes, I didn't mean `response` as a specific variable, just that I couldn't see how exported values would be accessed.  I guess I was expecting a variable exported by a named section to belong to that name, rather than to the global namespace. Maybe it would be helpful to have an "exported variables" section [Variables](https://httpyac.github.io/guide/variables.html) that mentions exports, which would then tie in to the [Access to Variables](https://httpyac.github.io/guide/scripting.html#access-to-variables) section on the scripting page.

**Comment by AnWeber** [comment]
I have already inserted the links only the style is unfortunately too unimpressive.
![image](https://user-images.githubusercontent.com/13434739/140195269-eaa3e08d-d371-4313-842d-304ba61f15e0.png)

You were actually the first to point out the principle with the scope of the variables. Before that, I had already noticed the erroneous behavior, but I had not yet paid such close attention to it. But an improvement would probably be to make the variables in VSCode more visible. This would make it easier to understand the behavior of the extension.
